src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/ServiceException.java

```diff
@@ -65,6 +68,28 @@ public ServiceException(String msg, ErrorDeviation errorDeviation) {
         super(msg, errorDeviation);
     }
 
-
-
+    /**
+     * Throws the caught ServiceException as a {@link DeviationAwareRuntimeException}. The cause and
+     * {@link ErrorDeviation#metadata ErrorDeviation metadata} will be transferred from the original exception
+     * but a new error code must be provided. If the underlying exception does not have error metadata then the
+     * localized message of the underlying exception is used as metadata instead. If there is no underlying exception
+     * at all, then only the error code will be returned within the ErrorDeviation object.
+     * @param newErrorCode the new error code of the exception
+     */
+    public void throwAsDeviationAwareRuntimeException(String newErrorCode) {
```

Comment:
Not doing a full review now, but just noticed this and wanted to add my thoughts.

This raises a few questions for me:
- many DeviationAwareExceptions offer constructors that make it possible to copy deviations from another exception (https://github.com/nordic-institute/X-Road/blob/8b7f2a6f789997c822dccd6120bb1ce4f3777f84/src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/openapi/BadRequestException.java#L42-L44). I thought it would maybe better to consider putting this kind of functionality to a DeviationAwareRuntimeException constructor instead. That way "deviation copying" theme would always be handled using the same style (target exception / it's constructors) and not sometimes by target exception, sometimes by source exception. Granted there are some problems with that option (lots of constructors)
- storing localizedMessage to metadata is a bit questionable. Metadata's reason of existence was originally to transfer some structured metadata to frontend: https://confluence.niis.org/pages/viewpage.action?pageId=8520435#APIGuidelines&Examples-Errorhandling. As in "duplicateService.v1" in the example. I know we are already breaking this (for example CodedException handling in exceptiontranslator). I fear adding a generic ServiceException.throwAsDeviationAwareRuntimeException will lead into wide use to easily and quickly translate ServiceExceptions into unchecked ones, and this creates an end result where metadata loses it's original purpose, you don't really know what kind of data it contains, and it often leaks exception messages to frontend even if that was not the intention.

I think the API and error object design started from the idea that we want to hide things like the stacktrace, and original unprocessed error message, from the client. And sometimes add something extra, the error metadata, which should be limited and well known and documented in the API - when calling this endpoint you may receive this metadata.

Then some contradictory requirements started piling up - we want to display all the error detail the old system does, and we don't want to be killed by the complex error handling, and this has caused the metadata idea to suffer and rot a little bit and become fuzzy.

There's always a bit of contradiction here - on the other hand we want to "show same level of detail in error messages as the old UI" - and on the other hand we want to follow https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/REST_Security_Cheat_Sheet.md#error-handling. I don't think we can create a universal solution for this, we need to decide on it case by case (and consult @petkivim and @raits maybe more often on these), and usually one of the requirements has to compromise (either UI displays less error detail than the old, or we break OWASP recommendations).

I think I would e.g. feel better if

```
    public void generateInternalTlsKeyAndCertificate() {
        try {
            externalProcessRunner.execute(generateCertScriptPath, generateCertScriptArgs.split("\\s+"));
            restartXroadProxy();
        } catch (ProcessNotExecutableException | ProcessFailedException e) {
            e.throwAsDeviationAwareRuntimeException(KEY_CERT_GENERATION_FAILED);
        }
    }

```
would make it more explicit on how it transforms e.localizedMessage -> e.error.metadata. When thinking about just this code, the difference is small, but I am more worried about how other code will (mis)use ServiceException.throwAsDeviationAwareRuntimeException

Reply:
Discussion in Slack: https://niis.slack.com/archives/GGBHVCYSY/p1580295667011100

Reply:
The process output is no longer returned to the API caller. Instead it is logged so that one can easily find it from the logs with the `correlation-id` that is provided with the response headers. The response body in error cases now only include an error code and status.